### DISCUSSION TITLE
Usage of bazel for (embedded C + code generation)

### DISCUSSION THREAD

**Discussion by jason-sachs** [bazelbuild/bazel#21750] [Q&A]
Anyone out there using bazel for embedded C with code generation?

I am trying to see how viable bazel would be to replace an existing build system. At present we make use of Python with code generation, based on YAML files for defining configuration options. There are some concepts I'm not sure how to determine whether they are supported:

- reading YAML files --- I see that [JSON is supported in Starlark](https://bazel.build/rules/lib/core/json), but not sure how we would handle YAML files, except with maybe a conversion step where each YAML file is converted to an equivalent JSON file that depends on the YAML file. (We don't use any fancy special YAML features; it's just there as a convenient human-editable place to put JSON-serializable data.)
- Python --- some of our build process uses various Python scripts to process data. This isn't going to go away, because of library dependencies (e.g. jsonschema for validating data models against a schema, numpy/scipy for some numerical analysis, a few others), but I expect that like any other tool that could be called by bazel, we should be able to run Python. Is this the case?

**Comment by moroten** [comment]
Yes, at [Meroton](https://meroton.com) we have multiple clients using embedded C with code generation using Python scripts and other languages. As stated in https://bazel.build/about/intro:

> Bazel supports projects in multiple languages and builds outputs for multiple platforms.

Any part of the build graph can use a tool to translate any language into something else. Additionally, you might have restrictions where some code generators, compilers or other tools only can execute in certain environments, e.g. OS, GPU, your own custom hardware, node locked licenses etc., and that also fits into the build graph with remote execution. (Yes, our clients do all of that as well.)

Code generators, compilers and other tools are also built by Bazel, so all the above applies: Use any language and code generators to compile the tool, and do the compilation on certain platforms if needed.

**Comment by tpudlik** [comment]
We use Bazel to build embedded C++ (and a little bit of C) in Pigweed
<https://pigweed.dev/>. It's an open source project, so you can poke around
to see how we've done it.

On the code-generation front specifically, we generate embedded-friendly
C++ from proto files (general docs <https://pigweed.dev/pw_protobuf/>, Bazel
rules
<https://cs.opensource.google/pigweed/pigweed/+/main:pw_protobuf_compiler/pw_proto_library.bzl>).
We do this using our custom Python plugin for the proto compiler. I would
say this (integrating multiple languages within a build) is an area in
which Bazel particularly shines: the Python interpreter and our Python
plugin code are managed by Bazel as build inputs, just like the proto files
and regular C++ source and header files. Everything gets rebuilt if and
only if it has to. There are pre-existing build rules for Python (in
rules_python <https://github.com/bazelbuild/rules_python>), just like there
are for C/C++.
